test/functional/kernel/region/tests/test-kernel-region-sync.hpp

```diff
@@ -0,0 +1,96 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef __TEST_KERNEL_REGION_SYNC_HPP__
+#define __TEST_KERNEL_REGION_SYNC_HPP__
+
+#include "test-kernel-region-utils.hpp"
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+template <typename INDEX_TYPE, typename WORKING_RES, typename EXEC_POLICY>
+void KernelRegionSyncFunctionalTest(INDEX_TYPE first, INDEX_TYPE last)
+{
+  camp::resources::Resource host_res{camp::resources::Host()};
+  camp::resources::Resource work_res{WORKING_RES()};
+
+  const INDEX_TYPE N = last - first;
+  
+  INDEX_TYPE* work_array1;
+  INDEX_TYPE* work_array2;
+  INDEX_TYPE* work_array3;
+
+  INDEX_TYPE* check_array;
+
+  allocRegionTestData(N,
+                      work_res,
+                      &work_array1, &work_array2, &work_array3,
+                      host_res,
+                      &check_array);
+
+  work_res.memset( work_array1, 0, sizeof(INDEX_TYPE) * N );
+  work_res.memset( work_array2, 0, sizeof(INDEX_TYPE) * N );
+  work_res.memset( work_array3, 0, sizeof(INDEX_TYPE) * N );
+
+  host_res.memset( check_array, 0, sizeof(INDEX_TYPE) * N );
+
+  std::vector<INDEX_TYPE> idx_array(N);
+  std::iota(idx_array.begin(), idx_array.end(), first);
+  std::reverse(idx_array.begin(), idx_array.end());
+  RAJA::TypedListSegment<INDEX_TYPE> lseg(&idx_array[0], N,
+                                          work_res);
+
+  RAJA::TypedRangeSegment<INDEX_TYPE> rseg(first, last);
```

Comment:
It might be worth adding a comment to explain why this test requires the syncthreads - it took me a bit of staring and jumping back and forth to the policy definition.

Reply:
OK. Will do.  I'm also going to add an issue to adjust our interface for sync operations and expand what we support.

Reply:
@davidbeckingsale I just pushed that change.

Reply:
@davidbeckingsale my comment change dismissed your approval. Please re-approve when you have a chance. Thanks.